test/bigdecimal/test_bigmath.rb

```diff
@@ -53,6 +53,16 @@ def test_cos
     assert_in_delta(0.0, cos(PI(N) * BigDecimal("301.5"), N))
   end
 
+  def test_tan
+    assert_in_delta(0.0, tan(BigDecimal("0.0"), N))
+    assert_in_delta(0.0, tan(PI(N), N))
+    assert_in_delta(1.0, tan(PI(N) / 4, N))
+    assert_in_delta(Math.sqrt(3.0), tan(PI(N) / 3, N))
```

Comment:
Use `BigMath.sqrt(BigDecimal(3), N)` instead of `Math.sqrt(3.0)`.
Moreover, I guess we need to use `assert_equal` instead of `assert_in_delta` for all the assertions in this test method.

Reply:
Note that I think we must replace `assert_in_delta` with `assert_equal` in all the other test methods, but they are out of this PR's target.

Reply:
I want to check the `10*N`-digit case in addition to the `N`-digit case for π/3.  So could you add the following assertion?

```suggestion
    assert_in_delta(Math.sqrt(3.0), tan(PI(N) / 3, N))
    assert_equal(sqrt(BigDecimal(3), 10*N), tan(PI(10*N) / 3, 10*N))
```

Reply:
Hello @mrkn, thank you for your review :pray: 

I used `assert_in_delta` for your suggestion because the definition of `tan` (`sin / cos`) makes the two methods' precision sum:

``rb
require "bigdecimal/math"

BigMath.sin(BigDecimal("1"), 10)
# => 0.84147098480789650665250231788602089836677454e0

BigMath.cos(BigDecimal("1"), 10)
# => 0.540302305868139717400936607637646275593203e0

BigMath.sin(BigDecimal("1"), 10) / BigMath.cos(BigDecimal("1"), 10)
# => 0.1557407724654902230506974799967261008769125258846034219578498192680719427656331763435628519e1
```

Using `assert_equal` in this case fails as the result has a lot more precision than `N`.

Do you think we should change the `tan` definition to handle precision differently?

Reply:
> Do you think we should change the tan definition to handle precision differently?

We must produce at least `N` precise digits for `BigMath.tan(x, N)`, so the answer is yes.
Instead, I think we can use `assert_in_delta` with specifying `BigDecimal("1e-#{N+1}")` in the `delta` argument.

Reply:
Oops, I'm sorry that I didn't consider how `assert_in_delta` is implemented.
`assert_in_delta` convert the arguments to Float, so we cannot use this to compare BigDecimal values!!

We need to prepare the BigDecimal version of `assert_in_delta` at first.

Reply:
To compare at least `N` digits, the following assertion may be enough:

```
y = BigDecimal("1.73205080756887729352744634150587236694280525381038")
assert_equal(y, BigMath.tan(BigMath.sqrt(3), 50).round(50, :truncate))
```

Reply:
Hello @mrkn 
Thank you again for reviewing this PR.

I had a look a other tests on the gem and I found that we use `assert_in_delta` in most of them, including `cos` and `sin`.
I understand that this may not be exactly what we're supposed to use for more accurate testing. Although, do you think we should first add this missing method (I'm also planning on adding `acos` and `asin` right after) the way the rest of the gem was implemented?

I'll be happy to help in the same time on updating gem to `assert_in_delta` in a way that is suitable for BigDecimal.
Please let me know what you think.

Reply:
The reason why I'm sticking to a precise alternative of `assert_in_delta` is that your implementation of `tan` depends on `sin` and `cos`, and they are tested by `assert_in_delta`.

If `tan` and other new methods you plan won't depend on other methods tested by `assert_in_delta`, it can be merged.

Reply:
Thank you for this explanation and the link to the work on `test-unit`, I see better why it is important to wait for a prior work.
I think it's better to stick to a definition of `tan` based on `sin` and `cos` for the simplicity of the gem (of course I'm all ears to counter arguments).

How can I help on your current PR on `test-unit`? This is a bit blurry to me but I guess I have to start somewhere.

Reply:
Fixing `assert_in_delta` is almost done.  After the new version of `test-unit` is released, I will update the master branch of this repository.  And then, please rebase this branch and fix the test.

> I think it's better to stick to a definition of `tan` based on `sin` and `cos` for the simplicity of the gem (of course I'm all ears to counter arguments).

If you can implement `tan` as a series that converges faster than `sin` and `cos`, using the series is better.  If not, you can at least depend on only `sin` by applying basic trigonometric formula `cos(x) = 1 - sin(x)**2`.  Could you please take and compare benchmark results of both `sin/cos` and `sin/(1 - sin**2)` to decide which is better?

Reply:
Additionally, we can treat the case of $x = 0$ as the special case that just returns 0.